### PR TITLE
add api v2

### LINKED ISSUE #2: Add APIv2
This plugin is missing APIv2 support.

### PR DESCRIPTION
**Pull request by timogoebel**
Fixes #2.

This PR adds an API for Snapshot management to the plugin. The commit also contains some more fixes. A lot of changes were required to dry-up the code. In addition, all rubocop warnings were fixed. The fog extensions were moves to concerns as helpers in rails are just "layout helpers".

I tried to limit changes to the absolute minimum, but the change is rather large as I tried to move all logic to the models to have light controllers.

This also adds some basic tests for the controllers.

Let me know, what you think.

### CODE REVIEW

**Comment by timogoebel** [comment]
I found some more TODOs while further testing this.
- `SnapshotController` needs to properly check the permissions of the `Host` object
- `Api::V2::SnapshotController` needs to properly handle when a Snapshot cannot be found.
- We need more API tests.

**Comment by mdellweg** [comment]
First of all, thank you very much. This is a big one. I do not yet understand everything, but moving the logic to a common place is, as always, a good idea.

**Comment by timogoebel** [comment]
@mdellweg: Added fixes for the issues mentioned above. Let me know if you have any questions, happy to answer any questions you have.

**Comment by mdellweg** [comment]
@timogoebel Would you mind, breaking the pullrequest up in roughly three commits?
As i understand, they could be
* move logic from controller to model/library and from helper to concerns
* add APIv2 controller
* add controller tests

**Comment by timogoebel** [comment]
@mdellweg: Sure, done. However, I needed three more commits to group this into logical changes.

File: app/models/foreman_snapshot_management/snapshot.rb [hunk 1]
```diff
@@ -4,70 +4,125 @@ class Snapshot
...
+    attr_accessor :id, :raw_snapshot, :name, :description, :host_id, :parent
 
-    def self.add_snapshot_with_children(snapshots, host, vmware_snapshot)
-      snapshots[vmware_snapshot.ref] = new_from_vmware(host, vmware_snapshot)
-      vmware_snapshot.child_snapshots.each do |snap|
-        add_snapshot_with_children(snapshots, host, snap)
-      end
+    def self.root_for_host(host)
+      raw_snapshot = host.compute_resource.get_snapshots(host.uuid).first
+      new_from_vmware(host, raw_snapshot)
```
Comment (mdellweg) [inline app/models/foreman_snapshot_management/snapshot.rb]: There is a problem with these lines. On a host without a snapshot, `first` returns `nil`. `new_from_vmware` cannot handle this. At the moment i do not see, whether it's safe to just return `nil` immediately.
I guess, `all_for_host` must return an empty list in that case.
I also infer, you are using the plugin extensively, so none of your hosts is left without snapshots. ;)

**Comment by timogoebel** [comment]
@mdellweg: Thanks for spotting this mistake. I believe I have found a better way to handle this. `fog-vsphere` already handles recursion and finding snapshots a little more efficient than we do. I pushed a commit, that relies on the library to return all snapshots for a host so we don't have to do it here. Let me know, what you think.

**Comment by mdellweg** [comment]
I liked it anyway, but now it works on my staging machine, too.
Thanks a lot for the good work.

**Comment by timogoebel** [comment]
@mdellweg: Thanks for the review. Do you mind releasing a new version with these changes to rubygems.org so we can package a new version?

**Comment by mdellweg** [comment]
@timogoebel Will do after adding another small in-house improvement.

**Pull request merged**